### ISSUE TITLE
[Epic] Support Go Generics

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi#9143]
With Go 1.18, generics are now [a part of the Go language](https://tip.golang.org/doc/go1.18#generics). Generics bring a lot of potential to simplify our Pulumi Go SDK, make it more type-safe, and dramatically reduce the amount of generated code in providers.

```[tasklist]
## M96 (Phase 2 sdkgen follow-ups and get started with Phase 2 programgen)
- [ ] https://github.com/pulumi/pulumi/issues/14344
- [ ] https://github.com/pulumi/pulumi/issues/13962
- [ ] https://github.com/pulumi/pulumi/issues/13963
- [ ] Go sdkgen: Code review feedback follow-ups
- [ ] Go sdkgen: Support Go schema flag to customize generic module `x` name
- [ ] Begin https://github.com/pulumi/pulumi/issues/13913
```

```[tasklist]
## M97 (Finish Phase 2 programgen)
- [ ] https://github.com/pulumi/pulumi/issues/13913
```

```[tasklist]
## M98 (Rollout, Docs, Blog)
- [ ] PR/FAQ
- [ ] https://github.com/pulumi/ci-mgmt/issues/601
- [ ] Update docs
- [ ] Update templates
- [ ] Update examples
- [ ] Blog post
```

## Completed Work

```[tasklist]
## M89
- [x] [Design](https://github.com/pulumi/pulumi/discussions/13057)
```

```[tasklist]
## M91
- [x] Phase 1 prototype
```

```[tasklist]
## M92
- [ ] https://github.com/pulumi/pulumi/issues/13585
```

```[tasklist]
## M93
- [ ] https://github.com/pulumi/pulumi/issues/13586
- [ ] https://github.com/pulumi/pulumi/issues/13587
- [ ] https://github.com/pulumi/pulumi/issues/13864
```

```[tasklist]
## M94
- [ ] https://github.com/pulumi/pulumi/issues/13912
- [x] Pulumi AWS preview branch
- [ ] https://github.com/pulumi/pulumi-hugo/issues/3356
- [ ] https://github.com/pulumi/pulumi-terraform-bridge/issues/1391
```

**Comment by ringods** [comment]
I notice that support for Go Generics is cut from the 4.0 changes tracking. Can someone from Pulumi explain why this was decided that way?

**Comment by DeluxeOwl** [comment]
Any news on this?

**Comment by mikhailshilkov** [comment]
We are currently in prototyping phase. The goals are

- Sketch the design of our Go SDK with generics
- Understand the value of this change for our users and ourselves
- Make the release plan, i.e. how much of a breaking change it is and how we can ship it in the least disruptive way

There's no specific timeline right now, but we still have it penciled in for 4.0 (whenever that comes).

**Comment by AaronFriel** [comment]
Prior art, adding this for posterity as we get closer to a design doc.

* An exploration of what Output APIs with generics: https://github.com/pulumi/go-generics
* Another exploration of an alternative API: https://github.com/t0yv0/pulumi-go-generics
* A branch of Pulumi exploring implementation of generic Output[T]: https://github.com/pulumi/pulumi/commit/fc2684485619425635b183da3a6eecd519f32bab

**Comment by ringods** [comment]
@AaronFriel repo https://github.com/pulumi/go-generics doesn't seem to be a public one. Can you update the access rights?

**Comment by AaronFriel** [comment]
I don't think there's much there aside from @pgavlin's experimentation.

**Comment by AaronFriel** [comment]
The contents of that repo are @pgavlin's experimentation with some alternate output APIs. I think he owns the repo, if he wants to open it up. 

I don't think we'll put any user-facing libraries in that repo, though.

**Comment by abhinav** [comment]
Update:
We have a design proposal up at https://github.com/pulumi/pulumi/discussions/13057.

**Comment by abhinav** [comment]
Status update:

**Current status**:
There were no objections to #13057 so that's the plan we're executing on.
As of right now, I have a prototype of the SDK and code generation changes from Phase 1 in the RFC. These changes are on the abhinav/go-generics branch on pulumi/pulumi. I'm frequently force-pushing to this branch as I flesh out the implementation, add documentation, and tests.
This branch was able to successfully generate the SDKs for pulumi-aws, pulumi-awsx, and pulumi-aws-native.

## Plan

### Phase 1

ref: https://github.com/pulumi/pulumi/discussions/13057#phase-1-generics-rollout

#### pulumi/pulumi

- [ ] Move parts of sdk/go/pulumi into sdk/go/pulumi/internal. (This is necessary to avoid a cyclic dependency between pulumi and pulumix.) The moved components will be re-exported from the go/pulumi package to retain backwards compatibility.
    - Status: Prototyped on the branch.
- [ ] Implement the pulumix package providing type-safe generic APIs for inputs, outputs, apply, etc.—everything in the Phase 1 section of the RFC.
    - Status: Prototyped on the branch. Most APIs implemented.
- [ ] Update code generation to implement the new generic `Input[T]` interface. This is backwards compatible.
    - Status: Prototyped on the branch.

In short, everything necessary for Phase 1 has been prototyped at this point. We're confident that this all fits together now, so we can begin turning this into production-ready code.

The above changes will be pushed to a branch `feature/go-generics`, which we'll maintain until these changes merge into master and are released.

#### Providers

##### Experimentation

To experiment with the changes above, given `$commitHash` of the HEAD of the feature branch, providers will need to make the following changes:

```
commitHash=  # set to commit hash
( cd sdk &&
    go mod edit -replace github.com/pulumi/pulumi/sdk/v3=github.com/pulumi/pulumi/sdk/v3@$commitHash &&
    go mod tidy -go=1.19 )
( cd provider &&
    go mod edit -replace github.com/pulumi/pulumi/sdk/v3=github.com/pulumi/pulumi/sdk/v3@$commitHash &&
    go mod tidy -go=1.19 &&
    go mod edit -replace github.com/pulumi/pulumi/pkg/v3=github.com/pulumi/pulumi/pkg/v3@$commitHash &&
    go mod tidy -go=1.19 )

# Regenerate code using commands like:
make tfgen build_go      # bridged provider, e.g. aws
make codegen generate_go # aws-native
```

##### Production

Once a new release of pulumi/pulumi is tagged with the aforementioned changes, providers should:

1. Update to the latest release of pulumi/pulumi
2. Regenerate their code
3. Release a new version

That's all that is needed from providers to be used with the generic APIs.

### Phase 2

ref: https://github.com/pulumi/pulumi/discussions/13057#phase-2-provider-migration

This cannot begin until all supported providers have released an SDK generated with the changes in Phase 1. This phase has the following steps that haven't begun yet:

- [ ] codegen/go: Add an opt-in flag that uses `pulumix`'s generic output types (`Input`, `Output`, `[G]{Ptr, Array, Map}[Output]`, etc.) instead of built-in primitive outputs (`StringOutput`, `IntOutput`, etc.) or generating container output types (no more `FooMapOutput`). Struct output wrappers should continue to be generated so that accessors are available (`FooOutput.Bar() Output[string]`).
- [ ] Opt providers into this flag incrementally as they become ready to tag new major versions—the suggested codegen changes are breaking changes.

**Comment by abhinav** [comment]
Based on recent discussion with @justinvp, note that the RFC (#13057) does differentiate between programgen and sdkgen—bucketing it all under codegen.

However, the work necessary in Phase 2 for programgen vs sdkgen differs significantly. Especially because programgen may need to know (somehow) whether a specific SDK is opted into using pux types or legacy pu types.

**Comment by marco-m** [comment]
Hello @justinvp, @Zaid-Ajaj. Would you be so kind as to give a status update on the whole Go generics epic? I am following with trepidation and excitement :-)

**Comment by justinvp** [comment]
Hi @marco-m, the short update is that we ran into Go compiler performance issues, particularly around memory usage that was causing problems for customers who weren't even using generics. We've therefore paused rollout of Go generics for at least the next few months until we have identified a workaround or if we can make it more "pay-for-play" in terms of compiler memory usage.

**Comment by marco-m** [comment]
Thanks for the quick update @justinvp, appreciated!
I was wondering:
- Do you plan to retest on each new release of the Go compiler? With some optimism, the problem might just be fixed by a new Go release :-)
- Is there a way for you to provide reproducible feedback to the Go project? That might help in surfacing the problem.

**Comment by yann-soubeyrand** [comment]
Hello, I’m wondering if there are news on this topic?
